src/porepy/compositional/materials.py

```diff
@@ -0,0 +1,534 @@
+        ``__setattr__`` of this base class to any child class which is a data class.
+
+        """
+
+        if is_dataclass(cls):
+            if hasattr(cls, "__post_init__"):
+                if cls.__post_init__ != Constants.__post_init__:
+                    raise AttributeError(
+                        f"Data classes inheriting form {Constants} must not"
+                        + " have __post_init__ defined, but inherit it."
+                    )
+            else:
+                object.__setattr__(cls, "__post_init__", Constants.__post_init__)
+            if hasattr(cls, "__setattr__"):
+                if cls.__setattr__ != Constants.__setattr__:
+                    raise AttributeError(
+                        f"Data classes inheriting form {Constants} must not"
+                        + " have __setattr__ defined, but inherit it."
+                    )
+            else:
+                object.__setattr__(cls, "__setattr__", Constants.__setattr__)
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        """Custom attribute setter imitating dataclasses with frozen attributes, but
+        only disallowing the setting of attributes which correspond to a field defined
+        as part of the material constant framework.
+
+        Parameters:
+            name: The name of an attribute.
+            value: The new value to be set.
+
+        Raises:
+            FrozenInstanceError: If the user attempts to set any of the following
+                attributes:
+
+                - :attr:`name`
+                - :attr:`units`
+                - :attr:`constants_in_SI`
+                - and any field of material parameters indirectly defined in
+                  :attr:`SI_units`
+
+        """
+
+        frozen = ["name", "units", "constants_in_SI", "_initialized"] + list(
+            type(self).SI_units.keys()
+        )
+        if name in frozen and self._initialized:
+            raise FrozenInstanceError(
+                f"Cannot asign to field {name}. Names, units and material parameters"
+                + " are frozen once set."
+            )
+        else:
+            super().__setattr__(name, value)
+
+    def to_units(self: _Constants, units: pp.Units) -> _Constants:
+        """Utility to quickly convert constants to new units.
+
+        Parameters:
+            units: A new unit system. Note that ``units`` must cover all SI units
+                declared in :attr:`SI_units` of the constants (child) class.
+
+        Returns:
+            A new instance of of the data class using this method, with parameters
+            converted according to ``units``.
+
+        """
+        return type(self)(name=self.name, units=units, **self.constants_in_SI)
+
+
+_Constants = TypeVar("_Constants", bound=Constants)
+"""Type variable for Constants-like objects. Mainly used to type the units conversion
+of :class:`Constants` correctly."""
+
+
+@dataclass(kw_only=True, eq=False)
```

Comment:
Why is eq set to false here? And is there a reason why it's true by default in SolidConstants?

Reply:
No need currently. The `eq=False` for `FluidComponent` is in order to make it hashable, due to omitting `frozen=True`.
Hashability must be given because `FluidComponent` is used as a dict key at various points.

`SolidConstants` is not there yet, so to speak.
But I can add it if wished! There is no point in creating an equality overload based on field values, since the dataclass is meant for representing distinct solids.

Reply:
For future reference, the issues with hashability for (eq=True,frozen=False) is just how dataclasses works, see https://stackoverflow.com/a/52390734.

I made the SolidConstants eq=False, and wrote a short discussion of why this happens. Could you have a look, @vlipovac, correct/improve if you think needed, and then resolve?